File: src/stores/listenTogether/protocol.ts
/**
 * 一起听协议 DTO 类型定义
 * 与安卓端 ListenTogether protocol 包完全对齐
 */

export const LtChannels = {
  NETEASE: 'netease',
  // 保留用于桌面本地映射，但跨端共享队列会排除该频道
  QQ_MUSIC: 'qqMusic',
  BILIBILI: 'bilibili',
  YOUTUBE_MUSIC: 'youtubeMusic',
  LOCAL: 'local',
} as const

/** ExoPlayer-aligned repeat integers used on the wire */
export const LtRepeatMode = {
  OFF: 0,
  ONE: 1,
  ALL: 2,
} as const

export type LtRepeatModeValue = (typeof LtRepeatMode)[keyof typeof LtRepeatMode]

export interface ListenTogetherTrack {
  stableKey: string
  channelId: string
  audioId: string
  subAudioId?: string
  playlistContextId?: string
  mediaUri?: string
  streamUrl?: string
  name: string
  artist: string
  album?: string
  durationMs: number
  coverUrl?: string
}

export interface ListenTogetherRoomSettings {
  allowMemberControl: boolean
  autoPauseOnMemberChange: boolean
  shareAudioLinks: boolean
}

export interface ListenTogetherMember {
  userUuid: string
  nickname: string
  userId?: string
  role: string
  joinedAt: number
}

export interface ListenTogetherPlaybackState {
  state: 'playing' | 'paused' | string
  basePositionMs: number
  baseTimestampMs: number
  playbackRate: number
  /** ExoPlayer: 0=OFF 1=ONE 2=ALL */
  repeatMode?: number | null
  shuffleEnabled?: boolean | null
}

export interface ListenTogetherRoomState {
  roomId: string
  version: number
  schemaVersion: number
  controllerUserUuid?: string
  controllerUserId?: string
  controllerHeartbeatAt?: number
  settings: ListenTogetherRoomSettings
  members: ListenTogetherMember[]
  queue: ListenTogetherTrack[]
  currentIndex: number
  track?: ListenTogetherTrack
  playback: ListenTogetherPlaybackState
  controllerOfflineSince?: number
  roomStatus: string
  closedReason?: string
  updatedAt: number
}

export interface ListenTogetherCause {
  userUuid?: string
  userId?: string
  nickname?: string
  eventId?: string
  type?: string
}

export interface ListenTogetherEvent {
  type: string
  eventId?: string
  clientTimeMs?: number
  clientInstanceId?: string
  clientSequence?: number
  positionMs?: number
  currentIndex?: number
  nextIndex?: number
  track?: ListenTogetherTrack
  queue?: ListenTogetherTrack[]
  roomSettings?: ListenTogetherRoomSettings
  shouldPlay?: boolean
  state?: string
  /** PLAYBACK_MODE / REQUEST_PLAYBACK_MODE */
  repeatMode?: number
  shuffleEnabled?: boolean
  requestTrackStableKey?: string
  finishedTrackStableKey?: string
}

export interface ListenTogetherSocketEnvelope {
  type: string
  sessionId?: string
  userUuid?: string
  userId?: string
  nickname?: string
  role?: string
  autoPauseOnJoin?: boolean
  state?: ListenTogetherRoomState
  expectedPositionMs?: number
  nowMs?: number
  t?: number
  ok?: boolean
  /// 控制命令应答：服务端拒绝时才有 error，缺它会让被拒绝的控制静默生效
  /// applied.state/causedBy 对齐 Android ListenTogetherAppliedEvent：
  /// 自己发起的 REQUEST_* 被仲裁通过后要据此把权威状态应用到本地播放器
  result?: {
    ok: boolean
    error?: string
    applied?: {
      type: string
      roomId?: string
      version?: number
      state?: ListenTogetherRoomState
      expectedPositionMs?: number
      nowMs?: number
      causedBy?: ListenTogetherCause
    }
  }
  message?: string
  roomId?: string
  version?: number
  causedBy?: ListenTogetherCause
  track?: ListenTogetherTrack
  queue?: ListenTogetherTrack[]
  positionMs?: number
  currentIndex?: number
  requestTrackStableKey?: string
  shouldPlay?: boolean
  stateName?: string
  repeatMode?: number
  shuffleEnabled?: boolean
  clientTimeMs?: number
  clientInstanceId?: string
  clientSequence?: number
  requestSequence?: number
}

export interface ListenTogetherInitialSnapshot {
  queue: ListenTogetherTrack[]
  currentIndex: number
  track?: ListenTogetherTrack
  settings: ListenTogetherRoomSettings
  isPlaying: boolean
  positionMs: number
  repeatMode: number
  shuffleEnabled: boolean
}

export interface ListenTogetherRoomResponse {
  ok: boolean
  roomId?: string
  userUuid?: string
  userId?: string
  nickname?: string
  role?: string
  autoPauseOnJoin?: boolean
  token?: string
  state?: ListenTogetherRoomState
  wsUrl?: string
  error?: string
}

export interface ListenTogetherStateResponse {
  ok: boolean
  state?: ListenTogetherRoomState
  expectedPositionMs?: number
  serverNowMs?: number
  autoPauseOnJoin?: boolean
  error?: string
}

export type ConnectionState = 'disconnected' | 'connecting' | 'connected'
export type LtRole = 'controller' | 'listener'

/** 播放命令来源 */
export type PlaybackCommandSource = 'local' | 'remote_sync'

/** Desktop string mode <-> wire int (ExoPlayer) */
export function desktopRepeatToWire(mode: string | undefined | null): number {
  switch (mode) {
    case 'one':
      return LtRepeatMode.ONE
    case 'all':
      return LtRepeatMode.ALL
    default:
      return LtRepeatMode.OFF
  }
}

export function wireRepeatToDesktop(mode: number | null | undefined): 'off' | 'one' | 'all' | null {
  if (mode === null || mode === undefined || Number.isNaN(mode)) return null
  if (mode === LtRepeatMode.ONE) return 'one'
  if (mode === LtRepeatMode.ALL) return 'all'
  if (mode === LtRepeatMode.OFF) return 'off'
  return null
}

// 昵称/房间号校验（对齐 Android ListenTogetherNicknameValidation / RoomIdValidation）
// Android 昵称白名单仅数字/大小写字母/汉字, 长度 1..24, 不含连字符
export const LT_NICKNAME_MAX_LENGTH = 24
const ROOM_ID_REGEX = /^[ABCDEFGHJKLMNPQRSTUVWXYZ23456789]{6}$/

function isAllowedNicknameChar(cp: number): boolean {
  if (cp >= 0x30 && cp <= 0x39) return true // 0-9
  if (cp >= 0x41 && cp <= 0x5a) return true // A-Z
  if (cp >= 0x61 && cp <= 0x7a) return true // a-z
  // CJK 统一表意文字（含扩展 A/B 常见区段），近似 Android HAN script 判定
  if (cp >= 0x4e00 && cp <= 0x9fff) return true
  if (cp >= 0x3400 && cp <= 0x4dbf) return true
  if (cp >= 0x20000 && cp <= 0x2a6df) return true
  return false
}

/** 返回 true 表示昵称合法（对齐 Android，避免云同步到 Android 端被 sanitize 丢弃） */
export function isValidLtNickname(nickname: string): boolean {
  const normalized = nickname.trim()
  if (normalized.length < 1 || normalized.length > LT_NICKNAME_MAX_LENGTH) return false
  for (const ch of normalized) {
    const cp = ch.codePointAt(0)
    if (cp === undefined || !isAllowedNicknameChar(cp)) return false
  }
  return true
}

/** roomId 归一化：去空白并大写（对齐 Android normalizeListenTogetherRoomId） */
export function normalizeLtRoomId(value: string): string {
  return value.trim().toUpperCase()
}

/** 返回 true 表示房间号合法 */
export function isValidLtRoomId(roomId: string): boolean {
  return ROOM_ID_REGEX.test(normalizeLtRoomId(roomId))
}

File: src-tauri/src/listen_together/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtTrack {
    pub stable_key: String,
    pub channel_id: String,
    pub audio_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sub_audio_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub playlist_context_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media_uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_url: Option<String>,
    pub name: String,
    pub artist: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub album: Option<String>,
    #[serde(default)]
    pub duration_ms: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cover_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtRoomSettings {
    #[serde(default = "default_true")]
    pub allow_member_control: bool,
    #[serde(default = "default_true")]
    pub auto_pause_on_member_change: bool,
    #[serde(default = "default_true")]
    pub share_audio_links: bool,
}

fn default_true() -> bool {
    true
}

impl Default for LtRoomSettings {
    fn default() -> Self {
        Self {
            allow_member_control: true,
            auto_pause_on_member_change: true,
            share_audio_links: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtMember {
    #[serde(default)]
    pub user_uuid: String,
    #[serde(default)]
    pub nickname: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    pub role: String,
    pub joined_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtPlaybackState {
    #[serde(default = "default_paused")]
    pub state: String,
    #[serde(default)]
    pub base_position_ms: i64,
    #[serde(default)]
    pub base_timestamp_ms: i64,
    #[serde(default = "default_rate")]
    pub playback_rate: f64,
    // Align Android ListenTogetherPlaybackState / ExoPlayer:
    // REPEAT_MODE_OFF=0, ONE=1, ALL=2
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repeat_mode: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shuffle_enabled: Option<bool>,
}

fn default_paused() -> String {
    "paused".to_string()
}
fn default_rate() -> f64 {
    1.0
}

impl Default for LtPlaybackState {
    fn default() -> Self {
        Self {
            state: "paused".to_string(),
            base_position_ms: 0,
            base_timestamp_ms: 0,
            playback_rate: 1.0,
            repeat_mode: None,
            shuffle_enabled: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtRoomState {
    pub room_id: String,
    pub version: i64,
    #[serde(default = "default_schema")]
    pub schema_version: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub controller_user_uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub controller_user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub controller_heartbeat_at: Option<i64>,
    #[serde(default)]
    pub settings: LtRoomSettings,
    #[serde(default)]
    pub members: Vec<LtMember>,
    #[serde(default)]
    pub queue: Vec<LtTrack>,
    #[serde(default)]
    pub current_index: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track: Option<LtTrack>,
    #[serde(default)]
    pub playback: LtPlaybackState,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub controller_offline_since: Option<i64>,
    #[serde(default = "default_active")]
    pub room_status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub closed_reason: Option<String>,
    #[serde(default)]
    pub updated_at: i64,
}

fn default_schema() -> i32 {
    1
}
fn default_active() -> String {
    "active".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtCause {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nickname: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtInitialSnapshot {
    #[serde(default)]
    pub queue: Vec<LtTrack>,
    #[serde(default)]
    pub current_index: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track: Option<LtTrack>,
    #[serde(default)]
    pub settings: LtRoomSettings,
    #[serde(default)]
    pub is_playing: bool,
    #[serde(default)]
    pub position_ms: i64,
    // Align Android ListenTogetherInitialSnapshot
    #[serde(default)]
    pub repeat_mode: i32,
    #[serde(default)]
    pub shuffle_enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtCreateRoomRequest {
    pub user_uuid: String,
    pub nickname: String,
    pub initial_snapshot: LtInitialSnapshot,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtJoinRoomRequest {
    pub user_uuid: String,
    pub nickname: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtRoomResponse {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub room_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nickname: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(default)]
    pub auto_pause_on_join: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<LtRoomState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ws_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtStateResponse {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<LtRoomState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expected_position_ms: Option<i64>,
    // Align Android ListenTogetherStateResponse.serverNowMs
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub server_now_ms: Option<i64>,
    #[serde(default)]
    pub auto_pause_on_join: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtEvent {
    pub r#type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_time_ms: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub client_instance_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub client_sequence: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_index: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_index: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track: Option<LtTrack>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub queue: Option<Vec<LtTrack>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub room_settings: Option<LtRoomSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub should_play: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<String>,
    // PLAYBACK_MODE / REQUEST_PLAYBACK_MODE payload (Android-aligned)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repeat_mode: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shuffle_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_track_stable_key: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub finished_track_stable_key: Option<String>,
}

/// 服务端实际落地的事件（对齐 Android ListenTogetherAppliedEvent）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtAppliedEvent {
    pub r#type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub room_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub state: Option<LtRoomState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_position_ms: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub now_ms: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub caused_by: Option<LtCause>,
}

/// 控制命令的应答（对齐 Android ListenTogetherControlResponse）
///
/// 缺这个字段时，服务端拒绝控制（成员无权限、房间已关闭等）在桌面端
/// 会被静默丢弃：本地乐观改动照旧生效，用户也收不到任何提示，
/// 于是两端状态就此分叉。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtControlResponse {
    pub ok: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub applied: Option<LtAppliedEvent>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LtSocketEnvelope {
    pub r#type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nickname: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(default)]
    pub auto_pause_on_join: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<LtRoomState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expected_position_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub now_ms: Option<i64>,
    // Android envelope also carries short alias `t`
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub t: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ok: Option<bool>,
    /// 控制命令应答；serde 默认忽略未知字段，缺这一项会被静默丢弃
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<LtControlResponse>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub room_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub caused_by: Option<LtCause>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track: Option<LtTrack>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub queue: Option<Vec<LtTrack>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_index: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_track_stable_key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub should_play: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repeat_mode: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shuffle_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_time_ms: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub client_instance_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub client_sequence: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_sequence: Option<i64>,
}
